compiler/noirc_frontend/src/elaborator/enums.rs

```diff
@@ -243,4 +253,672 @@ impl Elaborator<'_> {
             (pattern, parameter_type, Visibility::Private)
         }))
     }
+
+    /// To elaborate the rules of a match we need to go through the pattern first to define all
+    /// the variables within, then compile the corresponding branch. For each branch we do this
+    /// way we'll need to keep a distinct scope so that branches cannot access the pattern
+    /// variables from other branches.
+    ///
+    /// Returns (rows, result type) where rows is a pattern matrix used to compile the
+    /// match into a decision tree.
+    pub(super) fn elaborate_match_rules(
+        &mut self,
+        variable_to_match: DefinitionId,
+        rules: Vec<(Expression, Expression)>,
+    ) -> (Vec<Row>, Type) {
+        let result_type = self.interner.next_type_variable();
+        let expected_pattern_type = self.interner.definition_type(variable_to_match);
+
+        let rows = vecmap(rules, |(pattern, branch)| {
+            self.push_scope();
+            let pattern = self.expression_to_pattern(pattern, &expected_pattern_type);
+            let columns = vec![Column::new(variable_to_match, pattern)];
+
+            let guard = None;
+            let body_span = branch.span;
+            let (body, body_type) = self.elaborate_expression(branch);
+
+            self.unify(&body_type, &result_type, || TypeCheckError::TypeMismatch {
+                expected_typ: result_type.to_string(),
+                expr_typ: body_type.to_string(),
+                expr_span: body_span,
+            });
+
+            self.pop_scope();
+            Row::new(columns, guard, body)
+        });
+        (rows, result_type)
+    }
+
+    /// Convert an expression into a Pattern, defining any variables within.
+    fn expression_to_pattern(&mut self, expression: Expression, expected_type: &Type) -> Pattern {
+        let expr_span = expression.span;
+        let unify_with_expected_type = |this: &mut Self, actual| {
+            this.unify(actual, expected_type, || TypeCheckError::TypeMismatch {
+                expected_typ: expected_type.to_string(),
+                expr_typ: actual.to_string(),
+                expr_span,
+            });
+        };
+
+        match expression.kind {
+            ExpressionKind::Literal(Literal::Integer(x, negative)) => {
```

Comment:
Would it be easy to also support bools? Of course it could be a follow-up thing.

Reply:
Ah, yes. I meant to add a test case for bools today - seems like if I did it would've ran into me forgetting the bool case here. I'll add it - we should already support bools in the match compiler.